File: 98. ImplementTriePrefixTree.c
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>

typedef struct Trie {
    struct Trie *child[26];
    bool end;
} Trie;

Trie *trieCreate(void) {
    return calloc(1, sizeof(Trie));
}

void trieInsert(Trie *root, const char *word) {
    for (int i = 0; word[i]; ++i) {
        int idx = word[i] - 'a';
        if (!root->child[idx]) root->child[idx] = trieCreate();
        root = root->child[idx];
    }
    root->end = true;
}

bool trieSearch(Trie *root, const char *word) {
    for (int i = 0; word[i]; ++i) {
        int idx = word[i] - 'a';
        if (!root->child[idx]) return false;
        root = root->child[idx];
    }
    return root->end;
}

bool trieStartsWith(Trie *root, const char *prefix) {
    for (int i = 0; prefix[i]; ++i) {
        int idx = prefix[i] - 'a';
        if (!root->child[idx]) return false;
        root = root->child[idx];
    }
    return true;
}

int main(void) {
    Trie *trie = trieCreate();
    trieInsert(trie, "apple");
    puts(trieSearch(trie, "apple") ? "true" : "false");
    puts(trieSearch(trie, "app") ? "true" : "false");
    puts(trieStartsWith(trie, "app") ? "true" : "false");
    return 0;
}

File: 98. ImplementTriePrefixTree.java
public class implementTriePrefixTree {

    static class Trie {
        private final TrieNode root;

        Trie() {
            this.root = new TrieNode();
        }

        public void insert(String word) {
            TrieNode node = root;
            for (char ch : word.toCharArray()) {
                int index = ch - 'a';
                if (node.children[index] == null) {
                    node.children[index] = new TrieNode();
                }
                node = node.children[index];
            }
            node.isEndOfWord = true;
        }

        public boolean search(String word) {
            TrieNode node = traverse(word);
            return node != null && node.isEndOfWord;
        }

        public boolean startsWith(String prefix) {
            return traverse(prefix) != null;
        }

        private TrieNode traverse(String text) {
            TrieNode node = root;
            for (char ch : text.toCharArray()) {
                int index = ch - 'a';
                if (node.children[index] == null) {
                    return null;
                }
                node = node.children[index];
            }
            return node;
        }
    }

    static class TrieNode {
        TrieNode[] children = new TrieNode[26];
        boolean isEndOfWord;
    }

    public static void main(String[] args) {
        Trie trie = new Trie();

        trie.insert("apple");
        System.out.println(trie.search("apple"));
        System.out.println(trie.search("app"));
        System.out.println(trie.startsWith("app"));
        trie.insert("app");
        System.out.println(trie.search("app"));
    }
}
